packages/adapters/cache/src/lib/caches/transactions.ts

```diff
@@ -104,21 +115,57 @@ export class TransactionsCache extends Cache {
     });
   }
 
-  public async storeTxData(txs: XTransfer[]): Promise<void> {
-    for (const tx of txs) {
+  public async storeTransfers(transfers: XTransfer[]): Promise<void> {
+    const nonceIncreasesByDomain: { [domain: string]: boolean } = {};
+    const highestNonceByDomain: { [domain: string]: number } = {};
+    const newXCalls: { [transferId: string]: string } = {};
+    for (let transfer of transfers) {
+      const existing = await this.getTransferByTransferId(transfer.transferId);
+      // TODO: Sanity check: no update needed if this transfer is same as the one already stored!
+
+      // TODO: Sanity check: make sure that the existing transfer's metadata is the same as the old one!
+
+      // Update the existing transfer with the data from the new one; this will collate the transfer across
+      // domains, since our cache is indexed by transferId.
+      transfer = existing ? { ...existing, ...transfer } : transfer;
+      const { xcall, execute, reconcile, transferId, nonce: _nonce, originDomain } = transfer;
+      const nonce = Number(_nonce);
+      const stringified = JSON.stringify(transfer);
+
       // set transaction data at domain field in hash, hset returns the number of field that were added
       // gte(1) => added, 0 => updated,
       // reference: https://redis.io/commands/hset
-      await this.data.hset(`${this.prefix}:${tx.originDomain}`, `${tx.nonce}:${tx.transferId}`, JSON.stringify(tx));
-      //move pointer to latest Nonce
-      const latestNonce = (await this.data.hget(`${this.prefix}:${tx.originDomain}`, "latestNonce")) ?? "0";
-      if (Number(tx.nonce) > parseInt(latestNonce)) {
-        //if this txns nonce is > the current pointer to latest nonce point to this one now
-        await this.data.hset(`${this.prefix}:${tx.originDomain}`, "latestNonce", tx.nonce);
-        await this.data.publish(StoreChannel.NewHighestNonce, tx.nonce.toString());
+      await this.data.hset(`${this.prefix}:transfers`, `${transferId}`, stringified);
+      if (xcall.transactionHash && !execute?.transactionHash && !reconcile?.transactionHash) {
+        // If xcall but no execute or reconcile, then it's possibly a new transfer.
+        newXCalls[transferId] = stringified;
+      } else if (execute?.transactionHash || reconcile?.transactionHash) {
+        // If execute (or reconcile), then it's a completed transfer. In case we've previously recorded
+        // the xcall as new in this batch, we need to remove it from the newXCalls list.
+        delete newXCalls[transferId];
```

Comment:
in slow liquidity, you will still need to `execute` the transfer if it is `reconcile`d. i'm assuming this is just for sending bids, but if its for maintaining the cache for any future actions then this could be problematic

Reply:
oh you right, that comment is misleading. I should probs say "then it's a stale xcall"